Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/Internal/ThreadSafeMongoMobile.swift

```diff
@@ -45,110 +48,166 @@ class ThreadSafeMongoDatabase {
         self.databaseName = databaseName
         self.name = name
     }
 
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        try container.encode(databaseKey, forKey: .databaseKey)
+        try container.encode(dataDirectory, forKey: .dataDirectory)
+        try container.encode(name, forKey: .name)
+        try container.encode(databaseName, forKey: .databaseName)
+    }
+    required init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        self.databaseKey = try container.decode(String.self, forKey: .databaseKey)
+        self.databaseName = try container.decode(String.self, forKey: .databaseName)
+        self.name = try container.decode(String.self, forKey: .name)
+        self.dataDirectory = try container.decode(URL.self, forKey: .dataDirectory)
+    }
+
     fileprivate func underlyingCollection() throws -> MongoCollection<T> {
         return try CoreLocalMongoDBService
             .shared
-            .client(withClientAppID: clientAppID, withDataDirectory: dataDirectory)
+            .client(withInstanceKey: databaseKey, withDataDirectory: dataDirectory)
             .db(databaseName)
             .collection(name, withType: T.self)
     }
 
     func drop() throws {
-        try underlyingCollection().drop()
+        return try lock.write {
+            try underlyingCollection().drop()
+        }
     }
 
     func aggregate(_ pipeline: [Document], options: AggregateOptions? = nil) throws -> MongoCursor<Document> {
-        return try underlyingCollection().aggregate(pipeline, options: options)
+        return try lock.write {
+            return try underlyingCollection().aggregate(pipeline, options: options)
+        }
     }
 
     func count(_ filter: Document = Document(), options: CountOptions? = nil) throws -> Int {
-        return try underlyingCollection().count(filter, options: options)
+        return try lock.write {
+            return try underlyingCollection().count(filter, options: options)
+        }
     }
 
     func distinct(fieldName: String, filter: Document, options: DistinctOptions? = nil) throws -> [BSONValue?] {
-        return try underlyingCollection().distinct(fieldName: fieldName, filter: filter, options: options)
+        return try lock.write {
+            return try underlyingCollection().distinct(fieldName: fieldName, filter: filter, options: options)
+        }
     }
 
     func find() throws -> MongoCursor<T> {
-        return try underlyingCollection().find()
+        return try lock.write {
+            return try underlyingCollection().find()
+        }
     }
 
     func find(_ filter: Document, options: FindOptions? = nil) throws -> MongoCursor<T> {
-        return try underlyingCollection().find(filter, options: options)
+        return try lock.write {
+            return try underlyingCollection().find(filter, options: options)
+        }
     }
 
```

Comment:
Depending on which PR goes in first, I think my findOne() change will need to be editted